<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jelmer Poelstra">
<meta name="dcterms.date" content="2024-04-09">

<title>PRACS24: nf-core</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">pracs-sp24</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-week-overviews" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Week overviews</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-week-overviews">    
        <li>
    <a class="dropdown-item" href="../week0/overview.html">
 <span class="dropdown-text">Pre-course</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week1/w1_overview.html">
 <span class="dropdown-text">Week 1: Course intro, OSC and shell basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week2/w2_overview.html">
 <span class="dropdown-text">Week 2: Project organization, Markdown &amp; files in the shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week4/w4_overview.html">
 <span class="dropdown-text">Week 3: Git and GitHub</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week4/w4_overview.html">
 <span class="dropdown-text">Week 4: Shell script and CLI tools</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week5/w5_overview.html">
 <span class="dropdown-text">Week 5: Software and batch jobs at OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week6/w6_overview.html">
 <span class="dropdown-text">Week 6: Automated pipelines with Nextflow</span></a>
  </li>  
        <li class="dropdown-header">Week 7: Course recap (TBA)</li>
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures-i" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lectures I</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures-i">    
        <li class="dropdown-header">Week 1: Unix shell</li>
        <li>
    <a class="dropdown-item" href="../week1/w1_course-intro.html">
 <span class="dropdown-text">I: Course Intro (slides)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week1/w1_osc.html">
 <span class="dropdown-text">II: OSC Intro</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week1/w1_shell.html">
 <span class="dropdown-text">III: Shell basics</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Week 2: Project organization</li>
        <li>
    <a class="dropdown-item" href="../week2/w2_project-org.html">
 <span class="dropdown-text">I: Project organization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week2/w2_vscode-markdown.html">
 <span class="dropdown-text">II: VS Code &amp; Markdown</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week2/w2_shellfiles.html">
 <span class="dropdown-text">III: Managing files in the shell</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Week 3: Git</li>
        <li>
    <a class="dropdown-item" href="../week3/w3_git1.html">
 <span class="dropdown-text">I: Getting started with Git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week3/w3_git2.html">
 <span class="dropdown-text">II: Git remotes on GitHub</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week3/w3_git3.html">
 <span class="dropdown-text"><em>(Git bonus material)</em></span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Week 4: Shell scripts</li>
        <li>
    <a class="dropdown-item" href="../week4/w4_1_scripts.html">
 <span class="dropdown-text">I: Shell scripts</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week4/w4_2_cli-tools.html">
 <span class="dropdown-text">II: Running CLI tools with shell scripts</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week4/w4_3_bonus.html">
 <span class="dropdown-text"><em>(Scripts bonus material)</em></span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures-ii" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lectures II</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures-ii">    
        <li class="dropdown-header">Week 5: Software and batch jobs at OSC</li>
        <li>
    <a class="dropdown-item" href="../week5/w5_1_osc.html">
 <span class="dropdown-text">I: A closer look at OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week5/w5_2_software.html">
 <span class="dropdown-text">II: Using software at OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week5/w5_3_slurm.html">
 <span class="dropdown-text">III: Slurm batch jobs</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Week 6: Nexftlow</li>
        <li>
    <a class="dropdown-item" href="../week6/w6_1_pipelines.html">
 <span class="dropdown-text">I: Pipelines and workflow tools</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week6/w6_2_nfcore.html">
 <span class="dropdown-text">II: Running an nf-core Nextflow pipeline</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week6/w6_3_nextflow.html">
 <span class="dropdown-text">III: Intro to writing Nextflow pipelines</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Week 7: Recap</li>
        <li class="dropdown-header">Recap (TBA)</li>
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-homework" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Homework</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-homework">    
        <li class="dropdown-header">Exercises</li>
        <li>
    <a class="dropdown-item" href="../week1/w1_exercises.html">
 <span class="dropdown-text">Week 1: Shell basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week2/w2_exercises.html">
 <span class="dropdown-text">Week 2: Markdown and files in the shell</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week3/w3_exercises.html">
 <span class="dropdown-text">Week 3: Git and GitHub</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week4/w4_exercises.html">
 <span class="dropdown-text">Week 4: Shell scripts and CLI tools</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week5/w5_exercises.html">
 <span class="dropdown-text">Week 5: Software and batch jobs at OSC</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week6/w6_exercises.html">
 <span class="dropdown-text">Week 6: Nextflow</span></a>
  </li>  
        <li><hr class="dropdown-divider"></li>
        <li class="dropdown-header">Assignments</li>
        <li>
    <a class="dropdown-item" href="../week0/osc-setup.html">
 <span class="dropdown-text">Pre-course: OSC account setup</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../week2/w2_github-account.html">
 <span class="dropdown-text">Week 2: GitHub account</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-final-project" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Final project</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-final-project">    
        <li>
    <a class="dropdown-item" href="../finalproj/info.html">
 <span class="dropdown-text">General info</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../finalproj/proposal.html">
 <span class="dropdown-text">I: Proposal (Due Apr 8)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../finalproj/presentation.html">
 <span class="dropdown-text">II: Oral presentation (Apr 16 &amp; 18)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../finalproj/submission.html">
 <span class="dropdown-text">III: Final submission (Due Apr 29)</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-miscellaneous" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Miscellaneous</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-miscellaneous">    
        <li>
    <a class="dropdown-item" href="../about.html">
 <span class="dropdown-text">About this website</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ref/glossary.html">
 <span class="dropdown-text">Course glossary</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../ref/shell.html">
 <span class="dropdown-text">Unix shell reference</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="https://osu.instructure.com/courses/163340"> 
<span class="menu-text"><i class="fa-solid fa-graduation-cap" aria-label="graduation-cap"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://ondemand.osc.edu"> 
<span class="menu-text"><i class="fa-solid fa-desktop" aria-label="desktop"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://library.ohio-state.edu/record=b8624007~S7"> 
<span class="menu-text"><i class="fa-solid fa-book" aria-label="book"></i></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://library.ohio-state.edu/record=b9490023~S7"> 
<span class="menu-text"><i class="fa-solid fa-book-open" aria-label="book-open"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/pracs-sp24">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools-item" href="https://github.com/jelmerp/pracs-sp24/issues/new">
            Report an Issue
            </a>
          </li>
      </ul>
    </div>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#the-nf-core-rnaseq-pipeline" id="toc-the-nf-core-rnaseq-pipeline" class="nav-link active" data-scroll-target="#the-nf-core-rnaseq-pipeline"><span class="header-section-number">1</span> The nf-core rnaseq pipeline</a></li>
  <li><a href="#getting-set-up" id="toc-getting-set-up" class="nav-link" data-scroll-target="#getting-set-up"><span class="header-section-number">2</span> Getting set up</a>
  <ul class="collapse">
  <li><a href="#how-well-run-the-pipeline" id="toc-how-well-run-the-pipeline" class="nav-link" data-scroll-target="#how-well-run-the-pipeline"><span class="header-section-number">2.1</span> How we’ll run the pipeline</a></li>
  <li><a href="#activating-the-conda-environment" id="toc-activating-the-conda-environment" class="nav-link" data-scroll-target="#activating-the-conda-environment"><span class="header-section-number">2.2</span> Activating the Conda environment</a></li>
  <li><a href="#downloading-the-pipeline" id="toc-downloading-the-pipeline" class="nav-link" data-scroll-target="#downloading-the-pipeline"><span class="header-section-number">2.3</span> Downloading the pipeline</a></li>
  </ul></li>
  <li><a href="#writing-a-shell-script-to-run-the-pipeline" id="toc-writing-a-shell-script-to-run-the-pipeline" class="nav-link" data-scroll-target="#writing-a-shell-script-to-run-the-pipeline"><span class="header-section-number">3</span> Writing a shell script to run the pipeline</a>
  <ul class="collapse">
  <li><a href="#building-the-nextflow-run-command" id="toc-building-the-nextflow-run-command" class="nav-link" data-scroll-target="#building-the-nextflow-run-command"><span class="header-section-number">3.1</span> Building the <code>nextflow run</code> command</a></li>
  <li><a href="#creating-an-osc-configuration-file" id="toc-creating-an-osc-configuration-file" class="nav-link" data-scroll-target="#creating-an-osc-configuration-file"><span class="header-section-number">3.2</span> Creating an OSC configuration file</a></li>
  <li><a href="#the-final-script" id="toc-the-final-script" class="nav-link" data-scroll-target="#the-final-script"><span class="header-section-number">3.3</span> The final script</a></li>
  </ul></li>
  <li><a href="#running-the-pipeline" id="toc-running-the-pipeline" class="nav-link" data-scroll-target="#running-the-pipeline"><span class="header-section-number">4</span> Running the pipeline</a>
  <ul class="collapse">
  <li><a href="#preparing-the-sample-sheet" id="toc-preparing-the-sample-sheet" class="nav-link" data-scroll-target="#preparing-the-sample-sheet"><span class="header-section-number">4.1</span> Preparing the sample sheet</a></li>
  <li><a href="#submitting-your-shell-script" id="toc-submitting-your-shell-script" class="nav-link" data-scroll-target="#submitting-your-shell-script"><span class="header-section-number">4.2</span> Submitting your shell script</a></li>
  <li><a href="#checking-the-pipelines-progress" id="toc-checking-the-pipelines-progress" class="nav-link" data-scroll-target="#checking-the-pipelines-progress"><span class="header-section-number">4.3</span> Checking the pipeline’s progress</a></li>
  </ul></li>
  <li><a href="#the-pipelines-outputs" id="toc-the-pipelines-outputs" class="nav-link" data-scroll-target="#the-pipelines-outputs"><span class="header-section-number">5</span> The pipeline’s outputs</a></li>
  <li><a href="#self-study-material-a-closer-look-at-the-output" id="toc-self-study-material-a-closer-look-at-the-output" class="nav-link" data-scroll-target="#self-study-material-a-closer-look-at-the-output"><span class="header-section-number">6</span> Self-study material: A closer look at the output</a>
  <ul class="collapse">
  <li><a href="#the-multiqc-report" id="toc-the-multiqc-report" class="nav-link" data-scroll-target="#the-multiqc-report"><span class="header-section-number">6.1</span> The MultiQC report</a></li>
  <li><a href="#the-gene-count-table" id="toc-the-gene-count-table" class="nav-link" data-scroll-target="#the-gene-count-table"><span class="header-section-number">6.2</span> The gene count table</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Running an nf-core Nextflow pipeline</h1>
<p class="subtitle lead">Week 6 - part II</p>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jelmer Poelstra </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 9, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
This page is still under development
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<hr>
<p><br></p>
<section id="introduction" class="level4">
<h4 class="anchored" data-anchor-id="introduction">Introduction</h4>
<p>To learn how to run a Nextflow/nf-core pipeline, we’ll use <a href="https://nf-co.re/rnaseq">nf-core rnaseq</a>, which performs a reference-based RNA-seq analysis.</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="general-mechanics-of-running-a-pipeline" class="level4">
<h4 class="anchored" data-anchor-id="general-mechanics-of-running-a-pipeline">General mechanics of running a pipeline</h4>
<p>Running a pipeline like this is a bit different, and more involved, than running a typical piece of bioinformatics software — for example:</p>
<ul>
<li><p>The pipeline <strong>submits Slurm batch jobs for us</strong>, and tries to parallelize as much as possible, spawning <em>many</em> jobs. This also means we need a small “config file” to tell the pipeline how to submit jobs.</p></li>
<li><p>We need an <strong>installation</strong> of Nextflow, and a download of the pipeline files, while the pipeline runs all its constituent tools via separate Singularity containers that it will need to download first<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>.</p></li>
<li><p>Nextflow <strong>distinguishes between a final output dir and a “work dir”</strong>. All processes will run in various sub-dirs of the work dir, and after each process, output files are copied to the final output dir. Typically, not all outputs are copied, especially for intermediate steps like read trimming — and pipelines have settings to determine what is copied. This distinction is useful at OSC, where a Scratch dir is most suitable at the work dir (fast I/O and lots of storage space), and a Project dir is most suitable as the final output dir.</p></li>
<li><p>When you run a pipeline, there is a distinction between:</p>
<ul>
<li><strong>Pipeline-specific options</strong> to e.g.&nbsp;set inputs and outputs and customize what will be run (there can be 100+ of these for larger pipelines…). These by convention use a double dash <code>--</code>, e.g.&nbsp;<code>--input</code>.</li>
<li><strong>General Nextflow options</strong> to e.g.&nbsp;pass a configuration file for Slurm batch job submissions and determine resuming/rerunning behavior (see below). These always use a single dash, e.g.&nbsp;<code>-resume</code>.</li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="resuming-a-pipeline" class="level4">
<h4 class="anchored" data-anchor-id="resuming-a-pipeline">Resuming a pipeline</h4>
<p>Finally, we talked about the need for flexible rerunning of parts of a pipeline. Nextflow can do this when we use the <code>-resume</code> option, which will make a pipeline, for example:</p>
<ul>
<li>Start where it left off if the previous run failed before finishing, or timed out.</li>
<li>Only rerun what is needed after adding or removing samples</li>
<li>Similarly, will rerun only what is needed if we change/replace the reference genome files, which would be all steps that make use of these files and anything downstream of these steps.</li>
</ul>
<p><br></p>
</section>
<section id="the-nf-core-rnaseq-pipeline" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="the-nf-core-rnaseq-pipeline"><span class="header-section-number">1</span> The nf-core rnaseq pipeline</h2>
<p>The nf-core rnaseq pipeline is meant for RNA-seq projects that:</p>
<ul>
<li>Attempt to sequence only mRNA while avoiding non-coding RNAs (“mRNA-seq”).</li>
<li>Do not distinguish between RNA from different cell types (“bulk RNA-seq”).</li>
<li>Use short reads (≤150 bp) that do not cover full transcripts but do uniquely ID genes.</li>
<li>Use reference genomes (are reference-based) to associate reads with genes.</li>
<li>Downstream of this pipeline, such projects typically aim to statistically compare expression between groups of samples, and have multiple biological replicates per group.</li>
</ul>
<p>That might seem quite specific, but this is by far the most common use RNA-seq use case. The inputs of this pipeline are FASTQ files with raw reads, and reference genome files (assembly &amp; annotation), while the outputs include a gene count table and many “QC outputs”.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
What the pipeline does and does not do <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are typically two main parts to the kind of RNA-seq data analysis I just described, but this pipeline only does the first:</p>
<ul>
<li><strong>From reads to counts: yes</strong><br>
Generating a count table using the reads &amp; the reference genome.</li>
<li><strong>Count table analysis: no</strong><br>
Differential expression analysis, function enrichment analysis, etc.</li>
</ul>
<p>That makes sense because the latter part is not nearly as “standardized” as the first. It also does not need much computing power or parallelization, and is best done <em>interactively</em> using a language like R.</p>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
<section id="the-main-steps" class="level4">
<h4 class="anchored" data-anchor-id="the-main-steps">The main steps</h4>
<p>Let’s take a closer look at the steps in the pipeline:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/nfcore_rnaseq.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:100.0%"></p>
</figure>
</div>
<p><strong>Read QC and pre-processing</strong></p>
<ol type="1">
<li>Read QC (FastQC)</li>
<li>Adapter and quality trimming (TrimGalore)</li>
<li>Optional removal of rRNA (SortMeRNA) — off by default, but we will include this</li>
</ol>
<p><strong>Alignment &amp; quantification</strong></p>
<ol start="4" type="1">
<li>Alignment to the reference genome/transcriptome (STAR)</li>
<li>Gene expression quantification (Salmon)</li>
</ol>
<p><strong>Post-processing, QC, and reporting</strong></p>
<ol start="6" type="1">
<li>Post-processing alignments: sort, index, mark duplicates (samtools, Picard)</li>
<li>Alignment/count QC (RSeQC, Qualimap, dupRadar, Preseq, DESeq2)</li>
<li>Create a QC/metrics report (MultiQC)</li>
</ol>
<hr style="height:1pt; visibility:hidden;">
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Customizing what the pipeline runs <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>This pipeline is quite flexible and you can turn several steps off, add optional steps, and change individual options for most tools that the pipeline runs.</p>
<ul>
<li><strong>Optional removal of contaminants</strong> (BBSplit)<br>
Map to 1 or more additional genomes whose sequences may be present as contamination, and remove reads that map better to contaminant genomes.</li>
<li><strong>Alternative quantification routes</strong>
<ul>
<li>Use RSEM instead of Salmon to quantify.</li>
<li>Skip STAR and perform direct pseudo-alignment &amp; quantification with Salmon.</li>
</ul></li>
<li><strong>Transcript assembly and quantification (StringTie)</strong><br>
While the pipeline is focused on gene-level quantification, it does produce transcript-level counts as well (this is run by default).</li>
</ul>
</div>
</div>
</div>
<p><br></p>
</section>
</section>
<section id="getting-set-up" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="getting-set-up"><span class="header-section-number">2</span> Getting set up</h2>
<section id="how-well-run-the-pipeline" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="how-well-run-the-pipeline"><span class="header-section-number">2.1</span> How we’ll run the pipeline</h3>
<p>The entire pipeline can be run with a single command. But before we can do so, we’ll need to a bunch of prep, such as:</p>
<ul>
<li>Activating the software environment and downloading the pipeline files.</li>
<li>Defining the pipeline’s inputs and outputs, which includes creating a “sample sheet”.</li>
<li>Creating a small “config file” so Nextflow knows how to submit Slurm batch jobs at OSC.</li>
</ul>
<p>The main Nextflow process does not need much computing power (a single core with the default 4 GB of RAM will be sufficient) and even though our VS Code shell already runs on a compute and not a login node, we are still better off <strong>submitting the main process as a batch job as well</strong>, because:</p>
<ul>
<li>This process can run for many hours, and we don’t want to risk it disconnecting.</li>
<li>It’s good to store all the standard output about pipeline progress and so on in a file — this will automatically end up in a Slurm log file if we submit it as a batch job.</li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>Like we’ve done before, we will use both a <em>primary script</em> that will be submitted as a batch job, and a <em>runner script</em> with commands to run interactively including the submission of said batch job:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> week06/nfc-rnaseq</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> week06/nfc-rnaseq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> scripts run software</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">touch</span> scripts/nfc-rnaseq.sh run/run.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="activating-the-conda-environment" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="activating-the-conda-environment"><span class="header-section-number">2.2</span> Activating the Conda environment</h3>
<p>If you did <a href="../week5/w5_exercises.html">last week’s exercises</a>, you created a Conda environment with <a href="https://anaconda.org/bioconda/nextflow">Nextflow</a> and <a href="https://anaconda.org/bioconda/nf-core">nf-core tools</a>. If not, you can use my Conda environment. We will activate this environment in our runner script because we will first interactively download the pipeline files.</p>
<details>
<summary>
Code to create this Conda environment
</summary>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3/23.3.1-py310</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-y</span> <span class="at">-n</span> nextflow <span class="at">-c</span> bioconda nextflow=23.10.1 nf-core=2.13.1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this code into the run/run.sh script, then run it in the terminal]</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Activate the Conda environment (or use mine: /fs/ess/PAS0471/jelmer/conda/nextflow)</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3/23.3.1-py310</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate nextflow-23.10</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Check that Nextflow and nf-core tools can be run by printing the versions:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this code directly in the terminal]</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> <span class="at">-v</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>nextflow version 23.10.1.5891</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this code directly in the terminal]</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">nf-core</span> <span class="at">--version</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>                                          ,--./,-.
          ___     __   __   __   ___     /,-._.--~\
    |\ | |__  __ /  ` /  \ |__) |__         }  {
    | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                          `._,._,'

    nf-core/tools version 2.13.1 - https://nf-co.re

nf-core, version 2.13.1</code></pre>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="downloading-the-pipeline" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="downloading-the-pipeline"><span class="header-section-number">2.3</span> Downloading the pipeline</h3>
<p>We’ll use the <code>nf-core download</code> command to download the rnaseq pipeline’s files.</p>
<p>First, we need to set the environment variable <code>NXF_SINGULARITY_CACHEDIR</code> to tell Nextflow where to store the Singularity containers for all the tools the pipeline runs<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. We will use a dir of mine that already has all containers, to save some downloading time<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this code into the run/run.sh script, then run it in the terminal]</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an environment variable for the container dir</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_SINGULARITY_CACHEDIR</span><span class="op">=</span>/fs/ess/PAS0471/containers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Next, we’ll run the <code>nf-core download</code> command to download the currently latest version (<code>3.14.0</code>) of the rnaseq pipeline to <code>software/rnaseq</code>, and the associated container files to the previously specified dir:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this code into the run/run.sh script, then run it in the terminal]</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the nf-core rnaseq pipeline files</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">nf-core</span> download rnaseq <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--revision</span> 3.14.0 <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outdir</span> software/nfc-rnaseq <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--compress</span> none <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--container-system</span> singularity <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--container-cache-utilisation</span> amend <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--download-configuration</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/nfcore_download_output.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Explanation of all options given to <code>nf-core download</code> <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ul>
<li><code>--revision</code>: The version of the rnaseq pipeline.</li>
<li><code>--outdir</code>: The dir to save the pipeline definition files.</li>
<li><code>--compress</code>: Whether to compress the pipeline files — we chose not to.</li>
<li><code>--container-system</code>: The type of containers to download. This should always be <code>singularity</code> at OSC, because that’s the only supported type.</li>
<li><code>--container-cache-utilisation</code>: This is a little technical and not terribly interesting, but we used <code>amend</code>, which will make it check our <code>$NXF_SINGULARITY_CACHEDIR</code> dir for existing containers, and simply download any that aren’t already found there.</li>
<li><code>--download-configuration</code>: This will download some configuration files that we will actually not use, but if you don’t provide this option, it will ask you about it when you run the command.</li>
</ul>
<p>Also, don’t worry about the following warning, this doesn’t impact the downloading:</p>
<blockquote class="blockquote">
<p>WARNING Could not find GitHub authentication token. Some API requests may fail.</p>
</blockquote>
<hr style="height:1pt; visibility:hidden;">
</div>
</div>
</div>
<p>Let’s take a quick peek at the dirs and files we just downloaded:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this code directly in the terminal]</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> software/nfc-rnaseq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>3_14_0  configs</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this code directly in the terminal]</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> software/nfc-rnaseq/3_14_0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>assets        CODE_OF_CONDUCT.md  LICENSE       nextflow.config       subworkflows
bin           conf                main.nf       nextflow_schema.json  tower.yml
CHANGELOG.md  docs                modules       pyproject.toml        workflows
CITATIONS.md  lib                 modules.json  README.md</code></pre>
<p><br></p>
</section>
</section>
<section id="writing-a-shell-script-to-run-the-pipeline" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="writing-a-shell-script-to-run-the-pipeline"><span class="header-section-number">3</span> Writing a shell script to run the pipeline</h2>
<p>In this section, we’ll go through the components of the <code>scripts/nfc-rnaseq.sh</code> script that you’ll later submit as a Slurm batch job. The most important part of this script is the <code>nextflow run</code> command that will run the pipeline.</p>
<section id="building-the-nextflow-run-command" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="building-the-nextflow-run-command"><span class="header-section-number">3.1</span> Building the <code>nextflow run</code> command</h3>
<p>To run the pipeline, use the command <code>nextflow run</code>, followed by the path to the dir you just downloaded:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Partial shell script code, don't copy or run]</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run software/nfc-rnaseq/3_14_0</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>After that, there are several <strong>required options</strong> (see the <a href="https://nf-co.re/rnaseq/3.14.0/docs/usage">pipeline’s documentation</a>), which represent the input and output files/dirs for the pipeline:</p>
<ul>
<li><strong><code>--input</code></strong>: The path to a “sample sheet” with the paths to FASTQ files (more on that below).</li>
<li><strong><code>--fasta</code></strong>: The path to a reference genome assembly FASTA file — we’ll use the FASTA file we have in <code>data/ref</code>.</li>
<li><strong><code>--gtf</code></strong>: The path to a reference genome annotation file<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a> — we’ll use the GTF file we have in <code>data/ref</code>.</li>
<li><strong><code>--outdir</code></strong>: The path to the desired output dir for the <em>final</em> pipeline output — this can be whatever you like.</li>
</ul>
<p>This pipeline has different options for e.g.&nbsp;alignment and quantification. We will stick close to the defaults, which includes alignment with <code>STAR</code> and quantification with <code>Salmon</code>, with one exception: we want to remove reads from ribosomal RNA (this step is skipped by default).</p>
<hr style="height:1pt; visibility:hidden;">
<section id="exercise-finding-the-option-to-remove-rrna" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-finding-the-option-to-remove-rrna"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Finding the option to remove rRNA</h4>
<p>Take a look at the <a href="https://nf-co.re/rnaseq/parameters">“Parameters” tab on the pipeline’s documentation website</a>:</p>
<ul>
<li>Browse through the options for a bit to get a feel for the extent to which you can customize the pipeline.</li>
<li>Try to find the option to turn on removal of rRNA with SortMeRNA.</li>
</ul>
<details>
<summary>
<em>Click for the solution</em>
</summary>
The option we want is <strong><code>--remove_ribo_rna</code></strong>.
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<p>We’ll also use several <strong>general Nextflow options</strong> (note the single dash <code>-</code> notation; pipeline-specific options have <code>--</code>):</p>
<ul>
<li><strong><code>-profile</code></strong>: A so-called “profile” — should be <code>singularity</code> when running the pipeline with Singularity containers.</li>
<li><strong><code>-work-dir</code></strong>: The dir in which all the pipeline’s jobs/processes will run.</li>
<li><strong><code>-ansi-log false</code></strong>: Change Nextflow’s progress “logging” type to a format that works with Slurm log files<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>.</li>
<li><strong><code>-resume</code></strong>: Resume the pipeline where it “needs to” (e.g., where it left off) instead of always starting over.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
More on <code>-work-dir</code> and <code>-resume</code> <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><strong><code>work-dir</code></strong>:</p>
<p>The pipeline’s final outputs will go to the <code>--outdir</code> we talked about earlier. But all jobs/processes will run in, and initial outputs will be written to, a so-called <code>-work-dir</code>. After each process finishes, its key output files will then be copied to the final output dir. (There are also several pipeline options to customize what will and will not be copied.)</p>
<p>The distinction between such a work-dir and a final output dir can be very useful on HPC systems like OSC: you can use a scratch dir (at OSC: <code>/fs/scratch/</code>) with lots of storage space and fast I/O as the <code>work-dir</code>, and a backed-up project dir (at OSC: <code>/fs/ess/</code>) as the <code>outdir</code>, which will then not become unnecessarily large.</p>
<p><strong><code>-resume</code>:</strong></p>
<p>Besides resuming wherever the pipeline left off after an incomplete run (for example: it ran out of time or ran into an error), the <code>-resume</code> option also checks for any changes in input files or pipeline settings.</p>
<p>For example, if you have run the pipeline to completion previously, but rerun it after adding or replace one sample, <code>-resume</code> would make the pipeline <em>only</em> rerun the “single-sample steps” of the pipeline (which is most of them) for that sample as well as all steps that use all samples. Similarly, if you change an option that affects one of the first processes in the pipeline, the entire pipeline may be rerun, whereas if you change an option that only affects the last process, then only that last process would be rerun.</p>
<p>This option won’t make any difference when we run the pipeline for the first time, since there is nothing to resume. Nextflow will even give a warning along these lines, but this is not a problem.</p>
</div>
</div>
</div>
<hr style="height:1pt; visibility:hidden;">
<p>With all above-mentioned options, your final <code>nextflow run</code> command is:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Partial shell script code, don't copy or run]</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run software/nfc-rnaseq/3_14_0 <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input</span> <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fasta</span> <span class="st">"</span><span class="va">$fasta</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--gtf</span> <span class="st">"</span><span class="va">$gtf</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outdir</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--remove_ribo_rna</span> <span class="dt">\</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">-work-dir</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/raw <span class="dt">\</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">-profile</span> singularity <span class="dt">\</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">-ansi-log</span> false <span class="dt">\</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">-resume</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This uses several variables (e.g.&nbsp;<code>"$samplesheet"</code>) — these will enter the script via command-line arguments.</p>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="creating-an-osc-configuration-file" class="level3" data-number="3.2">
<h3 data-number="3.2" class="anchored" data-anchor-id="creating-an-osc-configuration-file"><span class="header-section-number">3.2</span> Creating an OSC configuration file</h3>
<p>We want the pipeline to submit Slurm batch jobs for us. We have to tell it to do this, and how, using a configuration (config) file. There are multiple ways of storing this file and telling Nextflow about it — the one we’ll use is to simply create a file <code>nextflow.config</code> in the dir from which we submit the <code>nextflow run</code> command: Nextflow will automatically detect and parse such a file.</p>
<p>We’ll keep this file as simple as possible, specifying only the “executor” program (in our case: Slurm) and the OSC project to use:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="st">process.executor = 'slurm'</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="st">process.clusterOptions='--account=PAS2700'</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="st">"</span> <span class="op">&gt;</span> nextflow.config</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="the-final-script" class="level3" data-number="3.3">
<h3 data-number="3.3" class="anchored" data-anchor-id="the-final-script"><span class="header-section-number">3.3</span> The final script</h3>
<p>Below is the full code for the script, in which I also added:</p>
<ul>
<li>Our standard shell script header lines.</li>
<li><code>#SBATCH</code> options: note that these are <strong>only</strong> for the “main” Nextflow job, not for the jobs that Nextflow itself will submit! So we ask for quite a bit of time<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>, but we don’t need more than the default 1 core and 4 GB of RAM.</li>
<li>Some <code>echo</code> reporting of arguments/variables, printing the date, etc.</li>
</ul>
<p>Open your <code>scripts/nfc-rnaseq.sh</code> script and paste the following into it:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --account=PAS2700</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=6:00:00</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mail-type=</span><span class="re">END</span><span class="co">,FAIL</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --output=slurm-nfc_rnaseq-%j.out</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Settings and constants</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="va">WORKFLOW_DIR</span><span class="op">=</span>software/nfc-rnaseq/3_14_0</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the Nextflow Conda environment</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load miniconda3/23.3.1-py310</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate /fs/ess/PAS0471/jelmer/conda/nextflow</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">NXF_SINGULARITY_CACHEDIR</span><span class="op">=</span>/fs/ess/PAS0471/containers</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Strict Bash settings</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="bu">set</span> <span class="at">-euo</span> pipefail</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Process command-line arguments</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a><span class="va">samplesheet</span><span class="op">=</span><span class="va">$1</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="va">fasta</span><span class="op">=</span><span class="va">$2</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span><span class="va">$3</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span><span class="va">$4</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Starting script nfc-rnaseq.sh"</span></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Samplesheet:          </span><span class="va">$samplesheet</span><span class="st">"</span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Reference FASTA:      </span><span class="va">$fasta</span><span class="st">"</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Reference GTF:        </span><span class="va">$gtf</span><span class="st">"</span></span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Output dir:           </span><span class="va">$outdir</span><span class="st">"</span></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span></span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the output dir</span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the config file</span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"</span></span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a><span class="st">process.executor = 'slurm'</span></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a><span class="st">process.clusterOptions='--account=PAS2700'</span></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a><span class="st">"</span> <span class="op">&gt;</span> nextflow.config</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the workflow</span></span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run <span class="st">"</span><span class="va">$WORKFLOW_DIR</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input</span> <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>    <span class="at">--fasta</span> <span class="st">"</span><span class="va">$fasta</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>    <span class="at">--gtf</span> <span class="st">"</span><span class="va">$gtf</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a>    <span class="at">--outdir</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span> <span class="dt">\</span></span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a>    <span class="at">--remove_ribo_rna</span> <span class="dt">\</span></span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a>    <span class="at">-work-dir</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/raw <span class="dt">\</span></span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>    <span class="at">-profile</span> singularity <span class="dt">\</span></span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a>    <span class="at">-ansi-log</span> false <span class="dt">\</span></span>
<span id="cb18-52"><a href="#cb18-52" aria-hidden="true" tabindex="-1"></a>    <span class="at">-resume</span></span>
<span id="cb18-53"><a href="#cb18-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-54"><a href="#cb18-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Report</span></span>
<span id="cb18-55"><a href="#cb18-55" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"Done with script nfc-rnaseq.sh"</span></span>
<span id="cb18-56"><a href="#cb18-56" aria-hidden="true" tabindex="-1"></a><span class="fu">date</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr style="height:1pt; visibility:hidden;">
<section id="exercise-take-a-close-look-at-the-script" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-take-a-close-look-at-the-script"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Take a close look at the script</h4>
<p>Go through your complete <code>scripts/nfc-rnaseq.sh</code> script and see if you understand everything that’s going on in there. Ask if you’re confused about anything!</p>
</section>
<p><br></p>
</section>
</section>
<section id="running-the-pipeline" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="running-the-pipeline"><span class="header-section-number">4</span> Running the pipeline</h2>
<p>We will now switch back to the <code>run/run.sh</code> script to add the code to submit the script. But we’ll have to create a sample sheet first.</p>
<section id="preparing-the-sample-sheet" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="preparing-the-sample-sheet"><span class="header-section-number">4.1</span> Preparing the sample sheet</h3>
<p>This pipeline requires a “sample sheet” as one of its inputs. In the sample sheet, you provide the paths to your FASTQ files and the so-called “strandedness” of your RNA-Seq library.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
RNA-Seq library strandedness <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>During RNA-Seq library prep, information about the directionality of the original RNA transcripts can be retained (resulting in a “stranded” library) or lost (resulting in an “unstranded” library: specify <code>unstranded</code> in the sample sheet).</p>
<p>In turn, stranded libraries can prepared either in reverse-stranded (<code>reverse</code>, by far the most common) or forward-stranded (<code>forward</code>) fashion. For more information about library strandedness, see <a href="https://www.azenta.com/blog/stranded-versus-non-stranded-rna-seq">this page</a>.</p>
<p>The pipeline also allows for a fourth option: <code>auto</code>, in which case the strandedness is automatically determined at the start of the pipeline by pseudo-mapping a small proportion of the data with Salmon.</p>
</div>
</div>
</div>
<p>The sample sheet should be a plain-text comma-separated values (CSV) file. Here is the example file from the <a href="https://nf-co.re/rnaseq/3.14.0/docs/usage#samplesheet-input">pipeline’s documentation</a>:</p>
<pre class="bash-out-solo"><code>sample,fastq_1,fastq_2,strandedness
CONTROL_REP1,AEG588A1_S1_L002_R1_001.fastq.gz,AEG588A1_S1_L002_R2_001.fastq.gz,auto
CONTROL_REP1,AEG588A1_S1_L003_R1_001.fastq.gz,AEG588A1_S1_L003_R2_001.fastq.gz,auto
CONTROL_REP1,AEG588A1_S1_L004_R1_001.fastq.gz,AEG588A1_S1_L004_R2_001.fastq.gz,auto</code></pre>
<p>So, we need a header row with column names, then one row per sample, and the following columns:</p>
<ul>
<li>Sample ID (we will simply use the file name part that is shared by R1 and R2).</li>
<li>R1 FASTQ file path (including the dir, unless these files are in your working dir).</li>
<li>R2 FASTQ file path (idem).</li>
<li>Strandedness: <code>unstranded</code>, <code>reverse</code>, <code>forward</code>, or <code>auto</code> — this data is forward-stranded, so we’ll use <code>forward</code>.</li>
</ul>
<p>You can create this file in several ways — we will do it here with a helper script that comes with the pipeline<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>:</p>
<ul>
<li><p>First, define an output dir (this will also be the output dir for the pipeline), and the sample sheet file name:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this into run/run.sh and then run it in the terminal]</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the output dir and sample sheet file name</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span>results/nfc-rnaseq</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="va">samplesheet</span><span class="op">=</span><span class="st">"</span><span class="va">$outdir</span><span class="st">"</span>/nfc_samplesheet.csv</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Next, run that helper script, specifying the strandedness of our data, the suffices of the R1 and R2 FASTQ files, and as arguments at the end, the input FASTQ dir (<code>data/fastq</code>) and the output file (<code>$samplesheet</code>):</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this into run/run.sh and then run it in the terminal]</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the sample sheet for the nf-core pipeline</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="ex">python3</span> software/nfc-rnaseq/3_14_0/bin/fastq_dir_to_samplesheet.py <span class="dt">\</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--strandedness</span> forward <span class="dt">\</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--read1_extension</span> <span class="st">"_R1.fastq.gz"</span> <span class="dt">\</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--read2_extension</span> <span class="st">"_R2.fastq.gz"</span> <span class="dt">\</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>    data/fastq <span class="dt">\</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Finally, check the contents of your newly created sample sheet file:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this directly in the terminal]</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>sample,fastq_1,fastq_2,strandedness
ERR10802863,data/fastq/ERR10802863_R1.fastq.gz,data/fastq/ERR10802863_R2.fastq.gz,forward
ERR10802864,data/fastq/ERR10802864_R1.fastq.gz,data/fastq/ERR10802864_R2.fastq.gz,forward
ERR10802865,data/fastq/ERR10802865_R1.fastq.gz,data/fastq/ERR10802865_R2.fastq.gz,forward
ERR10802866,data/fastq/ERR10802866_R1.fastq.gz,data/fastq/ERR10802866_R2.fastq.gz,forward
ERR10802867,data/fastq/ERR10802867_R1.fastq.gz,data/fastq/ERR10802867_R2.fastq.gz,forward
ERR10802868,data/fastq/ERR10802868_R1.fastq.gz,data/fastq/ERR10802868_R2.fastq.gz,forward
ERR10802869,data/fastq/ERR10802869_R1.fastq.gz,data/fastq/ERR10802869_R2.fastq.gz,forward
ERR10802870,data/fastq/ERR10802870_R1.fastq.gz,data/fastq/ERR10802870_R2.fastq.gz,forward
ERR10802871,data/fastq/ERR10802871_R1.fastq.gz,data/fastq/ERR10802871_R2.fastq.gz,forward
ERR10802874,data/fastq/ERR10802874_R1.fastq.gz,data/fastq/ERR10802874_R2.fastq.gz,forward
ERR10802875,data/fastq/ERR10802875_R1.fastq.gz,data/fastq/ERR10802875_R2.fastq.gz,forward
ERR10802876,data/fastq/ERR10802876_R1.fastq.gz,data/fastq/ERR10802876_R2.fastq.gz,forward
ERR10802877,data/fastq/ERR10802877_R1.fastq.gz,data/fastq/ERR10802877_R2.fastq.gz,forward
ERR10802878,data/fastq/ERR10802878_R1.fastq.gz,data/fastq/ERR10802878_R2.fastq.gz,forward
ERR10802879,data/fastq/ERR10802879_R1.fastq.gz,data/fastq/ERR10802879_R2.fastq.gz,forward
ERR10802880,data/fastq/ERR10802880_R1.fastq.gz,data/fastq/ERR10802880_R2.fastq.gz,forward
ERR10802881,data/fastq/ERR10802881_R1.fastq.gz,data/fastq/ERR10802881_R2.fastq.gz,forward
ERR10802882,data/fastq/ERR10802882_R1.fastq.gz,data/fastq/ERR10802882_R2.fastq.gz,forward
ERR10802883,data/fastq/ERR10802883_R1.fastq.gz,data/fastq/ERR10802883_R2.fastq.gz,forward
ERR10802884,data/fastq/ERR10802884_R1.fastq.gz,data/fastq/ERR10802884_R2.fastq.gz,forward
ERR10802885,data/fastq/ERR10802885_R1.fastq.gz,data/fastq/ERR10802885_R2.fastq.gz,forward
ERR10802886,data/fastq/ERR10802886_R1.fastq.gz,data/fastq/ERR10802886_R2.fastq.gz,forward</code></pre></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="submitting-your-shell-script" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="submitting-your-shell-script"><span class="header-section-number">4.2</span> Submitting your shell script</h3>
<p>As a last preparatory step, save the paths of the reference genome files in variables:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this into run/run.sh and then run it in the terminal]</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the reference genome files</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="va">fasta</span><span class="op">=</span>data/ref/GCF_016801865.2.fna</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="va">gtf</span><span class="op">=</span>data/ref/GCF_016801865.2.gtf</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Before you submit the script, check that all variables have been assigned by prefacing the command with <code>echo</code>:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [ Run this directly in the terminal]</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> sbatch scripts/nfc-rnaseq.sh <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span> <span class="st">"</span><span class="va">$fasta</span><span class="st">"</span> <span class="st">"</span><span class="va">$gtf</span><span class="st">"</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>sbatch scripts/nfc-rnaseq.sh results/nfc-rnaseq/nfc_samplesheet.csv data/ref/GCF_016801865.2.fna data/ref/GCF_016801865.2.gtf results/nfc-rnaseq</code></pre>
<p>If so, you are ready to submit the script as a batch job:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this into run/run.sh and then run it in the terminal]</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Submit the script to run the pipeline as a batch job</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="ex">sbatch</span> scripts/nfc-rnaseq.sh <span class="st">"</span><span class="va">$samplesheet</span><span class="st">"</span> <span class="st">"</span><span class="va">$fasta</span><span class="st">"</span> <span class="st">"</span><span class="va">$gtf</span><span class="st">"</span> <span class="st">"</span><span class="va">$outdir</span><span class="st">"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Submitted batch job 27767854</code></pre>
<hr style="height:1pt; visibility:hidden;">
<details>
<summary>
Not sure your <code>run.sh</code> script is complete, or getting errors? Click for its intended content.
</summary>
TODO
</details>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="checking-the-pipelines-progress" class="level3" data-number="4.3">
<h3 data-number="4.3" class="anchored" data-anchor-id="checking-the-pipelines-progress"><span class="header-section-number">4.3</span> Checking the pipeline’s progress</h3>
<p>Let’s check whether your job has started running, and if so, whether Nextflow has already spawned jobs:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Run this directly in the terminal]</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="ex">squeue</span> <span class="at">-u</span> <span class="va">$USER</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Mon Mar 25 12:13:38 2024
      JOBID PARTITION     NAME     USER    STATE   TIME TIME_LIMI  NODES NODELIST(REASON)
  27767854 serial-40 nfc-rnas   jelmer  RUNNING    1:33   6:00:00      1 p0219</code></pre>
<p>In the example output above, the only running job is the one we directly submitted, i.e.&nbsp;the main Nextflow process. The <code>NAME</code> column is the script’s name, <code>nfc-rnaseq.sh</code> (truncated to <code>nfc-rnas</code>).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
See examples of <code>squeue</code> output that includes Nextflow-submitted jobs <em>(Click to expand)</em>
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The top job, with partial name <code>nf-NFCOR</code>, is a job that’s been submitted by Nextflow:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="ex">squeue</span> <span class="at">-u</span> <span class="va">$USER</span> <span class="at">-l</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Mon Mar 25 13:14:53 2024
             JOBID PARTITION     NAME     USER    STATE       TIME TIME_LIMI  NODES NODELIST(REASON)
          27767861 serial-40 nf-NFCOR   jelmer  RUNNING       5:41  16:00:00      1 p0053
          27767854 serial-40 nfc_rnas   jelmer  RUNNING    1:03:48   6:00:00      1 p0219</code></pre>
<p>Unfortunately, the columns in the output above are quite narrow, so it’s <strong>not possible to see which step of the pipeline is being run by that job</strong>. The following (awful-looking!) code can be used to make that column much wider, so you can see the job’s full name which makes clear which step is being run (rRNA removal with SortMeRNA):</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="ex">squeue</span> <span class="at">-u</span> <span class="va">$USER</span> <span class="at">--format</span><span class="op">=</span><span class="st">"%.9i %.9P %.60j %.8T %.10M %.10l %.4C %R %.16V"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Mon Mar 25 13:15:05 2024
    JOBID PARTITION                                                          NAME    STATE       TIME TIME_LIMIT CPUS NODELIST(REASON)      SUBMIT_TIME
 27767861 serial-40   nf-NFCORE_RNASEQ_RNASEQ_SORTMERNA_(SRR27866691_SRR27866691)  RUNNING       5:55   16:00:00   12 p0053 2024-03-23T09:37
 27767854 serial-40                                                    nfc_rnaseq  RUNNING    1:04:02    6:00:00    1 p0219 2024-03-23T09:36</code></pre>
<p>You might also catch the pipeline while there are many more jobs running, e.g.:</p>
<pre class="bash-out-solo"><code>Mon Mar 25 13:59:50 2024
             JOBID PARTITION     NAME     USER    STATE       TIME TIME_LIMI  NODES NODELIST(REASON)
          27823107 serial-40 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0091
          27823112 serial-40 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0119
          27823115 serial-40 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0055
          27823120 serial-40 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0055
          27823070 serial-40 nf-NFCOR   jelmer  RUNNING       0:43  16:00:00      1 p0078
          27823004 serial-40 nfc-rnas   jelmer  RUNNING       2:13   6:00:00      1 p0146
          27823083 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0078
          27823084 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0096
          27823085 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0096
          27823086 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0115
          27823087 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0115
          27823088 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0123
          27823089 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0123
          27823090 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0057
          27823091 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0057
          27823092 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0058
          27823093 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0058
          27823095 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0118
          27823099 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0118
          27823103 serial-40 nf-NFCOR   jelmer  RUNNING       0:37  16:00:00      1 p0119
          27823121 serial-48 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0625
          27823122 serial-48 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0744
          27823123 serial-48 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0780
          27823124 serial-48 nf-NFCOR   jelmer  RUNNING       0:13  16:00:00      1 p0780</code></pre>
</div>
</div>
</div>
<p>You can also keep an eye on the pipeline’s progress, and see if there are any errors, by checking the Slurm log file — the top of the file should look like this:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># You will have a different job ID - replace as appropriate or use Tab completion</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> slurm-nfc_rnaseq-27767861.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>Starting script nfc-rnaseq.sh
Mon Mar 25 13:01:30 EDT 2024
Samplesheet:          results/nfc-rnaseq/nfc_samplesheet.csv
Reference FASTA:      data/ref/GCF_016801865.2.fna
Reference GTF:        data/ref/GCF_016801865.2.gtf
Output dir:           results/nfc-rnaseq

N E X T F L O W  ~  version 23.10.1
WARN: It appears you have never run this project before -- Option `-resume` is ignored
Launching `software/nfc-rnaseq/3_14_0/main.nf` [curious_linnaeus] DSL2 - revision: 746820de9b
WARN: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Multiple config files detected!
  Please provide pipeline parameters via the CLI or Nextflow '-params-file' option.
  Custom config files including those provided by the '-c' Nextflow option can be
  used to provide any configuration except for parameters.

  Docs: https://nf-co.re/usage/configuration#custom-configuration-files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


------------------------------------------------------
                                        ,--./,-.
        ___     __   __   __   ___     /,-._.--~'
  |\ | |__  __ /  ` /  \ |__) |__         }  {
  | \| |       \__, \__/ |  \ |___     \`-._,-`-,
                                        `._,._,'
  nf-core/rnaseq v3.14.0
------------------------------------------------------
Core Nextflow options
  runName                   : curious_linnaeus
  containerEngine           : singularity
[...output truncated...]</code></pre>
<p>The warnings about <code>-resume</code> and config files shown above can be ignored. Also, some of this output has nice colors that was not shown above, so here is a screenshot:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/slurmlog.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:95.0%"></p>
</figure>
</div>
<p>In the Slurm log file, pipeline progress is shown in the following way — you can only see which jobs are being submitted, not when they finish<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a>:</p>
<pre class="bash-out-solo"><code>[e5/da8328] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:PREPARE_GENOME:GTF_FILTER (GCF_016801865.2.fna)
[b5/9427a1] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:PREPARE_GENOME:CUSTOM_GETCHROMSIZES (GCF_016801865.2.fna)
[05/e0e09f] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802863)
[25/a6c2f5] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802863)
[24/cef9a0] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802864)
[b1/9cfa7e] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802864)
[c4/3107c1] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802865)
[7e/92ec89] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802866)
[01/f7ccfb] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802866)
[42/4b4da2] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802868)
[8c/fe6ca5] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:TRIMGALORE (ERR10802867)
[e6/a12ec8] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802867)
[2e/f9059d] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802865)
[de/2735d1] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:FASTQ_FASTQC_UMITOOLS_TRIMGALORE:FASTQC (ERR10802868)</code></pre>
<details>
<summary>
You should also see the following warning among the job submissions <em>(Click to expand)</em>
</summary>
<p>This warning can be ignored, the “Biotype QC” is not important and this information is indeed simply missing from our GTF file, there is nothing we can do about that.</p>
<pre class="bash-out-solo"><code>WARN: ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  Biotype attribute 'gene_biotype' not found in the last column of the GTF file!

  Biotype QC will be skipped to circumvent the issue below:
  https://github.com/nf-core/rnaseq/issues/460

  Amend '--featurecounts_group_type' to change this behaviour.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~</code></pre>
</details>
<hr style="height:1pt; visibility:hidden;">
<p>But if errors occur, they are reported in this file, and there is also a message when the entire pipeline has finished:</p>
<pre class="bash-out-solo"><code>[28/79e801] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_FORWARD:UCSC_BEDGRAPHTOBIGWIG (ERR10802864)
[e0/ba48c9] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_REVERSE:UCSC_BEDGRAPHTOBIGWIG (ERR10802864)
[62/4f8c0d] Submitted process &gt; NFCORE_RNASEQ:RNASEQ:MULTIQC (1)
-[nf-core/rnaseq] Pipeline completed successfully -
Done with script nfc-rnaseq.sh
Mon Mar 25 14:09:52 EDT 2024</code></pre>
<p><br></p>
</section>
</section>
<section id="the-pipelines-outputs" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="the-pipelines-outputs"><span class="header-section-number">5</span> The pipeline’s outputs</h2>
<p>You pipeline run may finish in as little as 15-30 minutes with our small test data set, but this can vary substantially, mostly due to variation in Slurm queue-ing times (the pipeline makes quite large resource requests!).</p>
<p>Once it has finished, you can take a look at the files and dirs in the specified output dir:</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-lh</span> results/nfc-rnaseq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>total 83K
drwxr-xr-x   2 jelmer PAS0471  16K Mar 25 13:02 fastqc
drwxr-xr-x   2 jelmer PAS0471 4.0K Mar 25 12:58 logs
drwxr-xr-x   3 jelmer PAS0471 4.0K Mar 25 13:14 multiqc
-rw-r--r--   1 jelmer PAS0471 2.0K Mar 25 19:55 nfc_samplesheet.csv
drwxr-xr-x   2 jelmer PAS0471 4.0K Mar 25 13:14 pipeline_info
drwxr-xr-x 248 jelmer PAS0471  16K Mar 25 13:10 raw
drwxr-xr-x   2 jelmer PAS0471 4.0K Mar 25 13:06 sortmerna
drwxr-xr-x  33 jelmer PAS0471  16K Mar 25 13:12 star_salmon
drwxr-xr-x   3 jelmer PAS0471 4.0K Mar 25 13:02 trimgalore</code></pre>
<p>The two most imporant outputs are:</p>
<ul>
<li><p>The <strong>MultiQC report</strong> (<code>&lt;outdir&gt;/multiqc/star_salmon/multiqc_report.html</code>): this has lots of QC summaries of the data, both the raw data and the alignments, and even a gene expression PCA plot.</p></li>
<li><p>The <strong>gene count table</strong> (<code>&lt;outdir&gt;/star_salmon/salmon.merged.gene_counts_length_scaled.tsv</code>): This is what you would use for downstream analysis such as differential expression and functional enrichment analysis.</p></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<p>To download the MultiQC HTML file at <code>results/nfc-rnaseq/multiqc/star_salmon/multiqc_report.html</code>, find this file in the VS Code explorer (file browser) on the left, right-click on it, and select <code>Download...</code>.</p>
<p>You can download it to any location on your computer. Then find the file on your computer and click on it to open it — it should be opened in your browser.</p>
<p>You can find a copy of the MultiQC report on this website, <a href="results/multiqc_report.html">here</a>. Go ahead and open that in a separate browser tab. There’s a lot of information in the report!</p>
<p><br></p>
</section>
<section id="self-study-material-a-closer-look-at-the-output" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="self-study-material-a-closer-look-at-the-output"><span class="header-section-number">6</span> Self-study material: A closer look at the output</h2>
<p>If you’re interested in RNA-seq data analysis, you may want to take a closer look at the outputs of the pipeline, especially MultiQC report.</p>
<section id="the-multiqc-report" class="level3" data-number="6.1">
<h3 data-number="6.1" class="anchored" data-anchor-id="the-multiqc-report"><span class="header-section-number">6.1</span> The MultiQC report</h3>
<p>Here are some items in that report to pay particular attention to, with example figures from this data set:</p>
<ul>
<li><p>The <strong>General Statistics</strong> table (the first section) is very useful, with the following notes:</p>
<ul>
<li><p>Most of the table’s content is also in later graphs, but the table allows for comparisons across metrics.</p></li>
<li><p>The <code>%rRNA</code> (% of reads identified as rRNA and removed by SortMeRNA) can only be found in this table.</p></li>
<li><p>It’s best to <em>hide the columns with statistics from Samtools</em>, which can be confusing if not downright misleading: click on “Configure Columns” and uncheck all the boxes for stats with Samtools in their name.</p></li>
<li><p>Some stats are for R1 and R2 files only, and some are for each sample as a whole. Unfortunately, this means you get 3 rows per sample in the table.</p></li>
</ul></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_table.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li>The <strong>Qualimap</strong> &gt; <strong>Genomic origin of reads</strong> plot shows, for each sample, the proportion of reads mapping to exonic vs.&nbsp;intronic vs.&nbsp;intergenic regions. This is an important QC plot: the vast majority of your reads should be exonic<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a>.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_genomic-origin.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption>This is a good result, with 80-90% of mapped reads in exonic regions.</figcaption>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li>The <strong>STAR</strong> &gt; <strong>Alignment Scores</strong> plot shows, for each sample, the percentage of reads that was mapped. Note that “Mapped to multiple loci” reads <em>are</em> also included in the final counts, and that “Unmapped: too short” merely means unmapped, really, and not that the reads were too short.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_star.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption>This is a pretty good results, with 80-90% of reads mapped.</figcaption>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li><strong>FastQC</strong> checks your FASTQ files, i.e.&nbsp;your data prior to alignment. There are FastQC plots both before and after trimming with TrimGalore/Cutadapt. The most important FastQC modules are:
<ul>
<li><strong>Sequence Quality Histograms</strong> — You’d like the mean qualities to stay in the “green area”.</li>
<li><strong>Per Sequence GC Content</strong> — Secondary peaks may indicate contamination.</li>
<li><strong>Adapter Content</strong> — Any adapter content should be gone in the post-trimming plot.</li>
</ul></li>
</ul>
<hr style="height:1pt; visibility:hidden;">
<section id="exercise-interpreting-fastqc-results-in-the-multiqc-report" class="level4 exercise">
<h4 class="anchored" data-anchor-id="exercise-interpreting-fastqc-results-in-the-multiqc-report"><i class="fa-solid fa-user-edit" aria-label="user-edit"></i> Exercise: Interpreting FastQC results in the MultiQC report</h4>
<p>Take a look at the three FastQC modules discussed above, both before and after trimming.</p>
<ul>
<li>Has the base quality improved after trimming, and does this look good?</li>
</ul>
<details>
<summary>
<em>Click to see the answer</em>
</summary>
<ul>
<li>Pre-trimming graph: The qualities are good overall, but there is more variation that what is usual, and note the poorer qualities in the first 7 or so bases. There is no substantial decline towards the end of the read as one often sees with Illumina data, but this is expected given that the reads are only 75 bp.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_basequal_pretrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Pre-trimming</strong> (Mean base quality scores: one line is one sample.)</figcaption>
</figure>
</div>
<ul>
<li>Post-trimming graph: The qualities have clearly improved. The first 7 or so bases remain of clearly poorer quality, on average.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_basequal_posttrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Post-trimming</strong></figcaption>
</figure>
</div>
</details>
<ul>
<li>Do you have any idea what’s going with the pre-trimming GC content distribution? What about after trimming — does this look good or is there reason to worry?</li>
</ul>
<details>
<summary>
<em>Click to see the answer</em>
</summary>
<ul>
<li>The pre-trimming GC content is very odd but this is mostly due to a high number of reads with zero and near-zero percent GC content. These are likely reads with only Ns. There are also some reads with near-hundred percent GC content. These are likely artifactual G-only reads that NextSeq/NovaSeq machines can produce.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_gc-pretrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Pre-trimming</strong>. One line is one file.</figcaption>
</figure>
</div>
<ul>
<li>After trimming, things look a lot better but there may be contamination here, given the weird “shoulder” at 30-40% GC.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_gc-posttrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Post-trimming</strong></figcaption>
</figure>
</div>
</details>
<ul>
<li>Do you know what the “adapters” that FastQC found pre-trimming are? Were these sequences removed by the trimming?</li>
</ul>
<details>
<summary>
<em>Click to see the answer</em>
</summary>
<ul>
<li>Pre-trimming, there seem to be some samples with very high adapter content throughout the read. This doesn’t make sense for true adapters, because these are usually only found towards the end of the read, when the read length is longer than the DNA fragment length. If you hover over the lines, you’ll see it says “polyg”. These are artifactual G-only reads that NextSeq/NovaSeq can produce, especially in the reverse reads — and you can see that all of the lines are for reverse-read files indeed.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_adapter-pretrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Pre-trimming</strong></figcaption>
</figure>
</div>
<ul>
<li>Post-trimming, no adapter content was found.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_adapter-posttrim.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:90.0%"></p>
<figcaption><strong>Post-trimming</strong></figcaption>
</figure>
</div>
</details>
</section>
<hr style="height:1pt; visibility:hidden;">
<ul>
<li>The <strong>Qualimap</strong> &gt; <strong>Gene Coverage Profile</strong> plot. This shows average read-depth across the position of genes/transcripts (for all genes together), which helps to assess the amount of RNA degradation. For poly-A selected libraries, RNA molecules “begin” at the 3’ end (right-hand side of the graph), so the more degradation there is, the more you expect there to be a higher read-depth towards the 3’ end compared to the 5’ end. (Though note that sharp decreases at the <em>very</em> end on each side are expected.)</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_gene-coverage.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:85.0%"></p>
<figcaption>There depth at ~20% (near the 5’ end) is clearly lower than at ~80% (near the 3’ end),<br>indicating some RNA degradation.</figcaption>
</figure>
</div>
<ul>
<li>The <strong>RSeqQC</strong> &gt; <strong>Infer experiment</strong> (library strandedness) plot. If your library is:
<ul>
<li>Unstranded, there should be similar percentages of Sense and Antisense reads.</li>
<li>Forward-stranded, the vast majority of reads should be Sense.</li>
<li>Reverse-stranded, the vast majority of reads should be Antisense.</li>
</ul></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_strandedness.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:85.0%"></p>
<figcaption>This libary is clearly forward-stranded, as we indicated in our sample sheet.</figcaption>
</figure>
</div>
<ul>
<li>The <strong>STAR_SALMON DESeq2 PCA</strong> plot is from a Principal Component Analysis (PCA) run on the final gene count table, thus showing overall patterns of gene expression similarity among samples.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/multiqc_pca.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:80.0%"></p>
<figcaption>The sampels clear form two distinct groups along PC1.</figcaption>
</figure>
</div>
<hr style="height:1pt; visibility:hidden;">
</section>
<section id="the-gene-count-table" class="level3" data-number="6.2">
<h3 data-number="6.2" class="anchored" data-anchor-id="the-gene-count-table"><span class="header-section-number">6.2</span> The gene count table</h3>
<p>The gene count table has one row for each gene and one column for each sample, with the first two columns being the <code>gene_id</code> and <code>gene_name</code><a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a>. Each cell’s value contains the read count estimate for a specific gene in a specific sample:</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># [Paste this into the run/run.sh script and run it in the terminal]</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the count table:</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># ('column -t' lines up columns, and less's '-S' option turns off line wrapping)</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="va">counts</span><span class="op">=</span>results/nfc-rnaseq/star_salmon/salmon.merged.gene_counts_length_scaled.tsv</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="ex">column</span> <span class="at">-t</span> <span class="st">"</span><span class="va">$counts</span><span class="st">"</span> <span class="kw">|</span> <span class="fu">less</span> <span class="at">-S</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre class="bash-out"><code>gene_id             gene_name           ERR10802863        ERR10802864        ERR10802865        ERR10802866        ERR10802867        ERR10802868       
ATP6                ATP6                163.611027228009   178.19903533081    82.1025390726658   307.649552934133   225.78249209207    171.251589309856  
ATP8                ATP8                0                  1.01047333891691   0                  0                  0                  0                 
COX1                COX1                1429.24769032452   2202.82009602881   764.584344577622   2273.6965332904    2784.47391614249   2000.51277019854  
COX2                COX2                116.537361366535   175.137972566817   54.0166352459629   256.592955351283   193.291937038438   164.125833130119  
COX3                COX3                872.88670991359    1178.29247734231   683.167933227141   1200.01735304529   1300.3853323715    1229.11746824104  
CYTB                CYTB                646.028108528182   968.256051104547   529.393909319439   1025.23768317788   1201.46662840336   842.533209911258  
LOC120412322        LOC120412322        0                  0                  0                  0                  0.995135178345792  0.996805450081561 
LOC120412324        LOC120412324        37.8326244586681   20.9489661184365   27.6702324729125   48.6417838830061   22.8313729348804   36.87899862428    
LOC120412325        LOC120412325        3.21074365394071   2.10702898851342   4.40315394778926   5.47978997387391   4.33241716734803   4.23386924919438  
LOC120412326        LOC120412326        0                  0                  0                  0                  0                  0                 
LOC120412327        LOC120412327        37.8206758601034   35.9063291323018   38.517771617566    27.7802608986967   37.6979028802121   32.885944667709   
LOC120412328        LOC120412328        35.0080600370267   20.0019192467143   23.9260736995594   30.0191332346116   21.0383665366408   28.9844776623531  
LOC120412329        LOC120412329        121.777922287929   112.794544755113   131.434181046282   127.753086659103   114.864750589664   131.589608063253  
LOC120412330        LOC120412330        42.8505448763697   28.9442284428204   36.6285174684674   46.7310765909945   42.7633834468768   26.9265243413636  
LOC120412331        LOC120412331        11.013179311581    9.00559907892481   12.9836833055803   13.029954361225    7.02624958751718   16.000552787954   
LOC120412332        LOC120412332        12.1055360835441   26.1231316926989   21.2767913384733   18.2783703626438   26.4932540325187   22.098808637857   
LOC120412333        LOC120412333        19.1159998132169   17.0558058070299   12.0965688236319   14.1510477997588   15.2033452089903   9.02624985028677  
LOC120412334        LOC120412334        9.01332125155807   3.00232591636489   5.99566364212933   11.0306919231504   8.03448732510427   11.0022053123759  
# [...output truncated...]</code></pre>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Count table versions
</div>
</div>
<div class="callout-body-container callout-body">
<p>The workflow outputs several versions of the count table<a href="#fn11" class="footnote-ref" id="fnref11" role="doc-noteref"><sup>11</sup></a>, but the one with <code>gene_counts_length_scaled</code> is the one we want:</p>
<ul>
<li><code>gene_counts</code> as opposed to <code>transcript_counts</code> for counts that are summed across transcripts for each gene.</li>
<li><code>length</code> for estimates that have been adjusted to account for between-sample differences in mean transcript length (longer transcripts would be expected to produce more reads in sequencing).</li>
<li><code>scaled</code> for estimates that have been scaled back using the “library sizes”, per-sample total read counts.</li>
</ul>
</div>
</div>
<p><br> <br></p>


</section>
</section>


<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a><div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p> Conda environments can also be used, but containers are recommended, and are what we will use.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p> These kinds of settings are more commonly specified with command <em>options</em>, but somewhat oddly, this is the only way we can specify that here.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p> But if you want to run a pipeline yourself for your own research, make sure to use a dir that you have permissions to write to.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p> Preferably in GTF (<code>.gtf</code>) format, but the pipeline can accept GFF/GFF3 (<code>.gff</code>/<code>.gff3</code>) format files as well.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p> The default logging does not work well the output goes to a text file, as it will in our case because we will submit the script with the Nextflow command as a Slurm batch job.<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p> And for a run of a full data set, you may want to ask even more, e.g.&nbsp;12-24 hours.<a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p> The box below shows an alternative method with Unix shell commands<a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p> The default Nextflow logging (without <code>-ansi-log false</code>) does show when jobs finish, but this would result in very messy output in a Slurm log file.<a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn9"><p> A lot of intronic content may indicate that you have a lot of pre-mRNA in your data; this is more common when your library prep used rRNA depletion instead of poly-A selection. A lot of intergenic content may indicate DNA contamination. Poor genome annotation quality may also contribute to a low percentage of exonic reads. The <em>RSeQC</em> &gt; <em>Read Distribution</em> plot will show this with even more categories, e.g.&nbsp;separately showing UTRs.<a href="#fnref9" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn10"><p> Which happen to be the same here, but these are usually different.<a href="#fnref10" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn11"><p> And each version in two formats: <code>.rds</code> (a binary R object file type) and <code>.tsv</code>.<a href="#fnref11" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>2024, Jelmer Poelstra</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/jelmerp/pracs-sp24">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>




</body></html>